rapids-cmake/cpm/bs_thread_pool.cmake

```diff
@@ -0,0 +1,131 @@
+
+  rapids_cpm_bs_thread_pool( [BUILD_EXPORT_SET <export-name>]
+                             [INSTALL_EXPORT_SET <export-name>]
+                             [<CPM_ARGS> ...])
+
+.. |PKG_NAME| replace:: bs_thread_pool
+.. include:: common_package_args.txt
+
+Result Targets
+^^^^^^^^^^^^^^
+  `BS::thread_pool` target will be created
+
+Result Variables
+^^^^^^^^^^^^^^^^
+  :cmake:variable:`bs_thread_pool_SOURCE_DIR` is set to the path to the source directory of `thread-pool`.
+  :cmake:variable:`bs_thread_pool_BINARY_DIR` is set to the path to the build directory of `thread-pool`.
+  :cmake:variable:`bs_thread_pool_ADDED`      is set to a true value if `thread-pool` has not been added before.
+  :cmake:variable:`bs_thread_pool_VERSION`    is set to the version of `thread-pool` specified by the versions.json.
+
+#]=======================================================================]
+function(rapids_cpm_bs_thread_pool)
+  list(APPEND CMAKE_MESSAGE_CONTEXT "rapids.cpm.bs_thread_pool")
+
+  set(options)
+  set(one_value BUILD_EXPORT_SET INSTALL_EXPORT_SET)
+  set(multi_value)
+  cmake_parse_arguments(_RAPIDS "${options}" "${one_value}" "${multi_value}" ${ARGN})
+
+  # Fix up _RAPIDS_UNPARSED_ARGUMENTS to have EXPORT_SETS as this is need for rapids_cpm_find
+  if(_RAPIDS_INSTALL_EXPORT_SET)
+    list(APPEND _RAPIDS_EXPORT_ARGUMENTS INSTALL_EXPORT_SET ${_RAPIDS_INSTALL_EXPORT_SET})
+  endif()
+  if(_RAPIDS_BUILD_EXPORT_SET)
+    list(APPEND _RAPIDS_EXPORT_ARGUMENTS BUILD_EXPORT_SET ${_RAPIDS_BUILD_EXPORT_SET})
+  endif()
+  set(_RAPIDS_UNPARSED_ARGUMENTS ${_RAPIDS_EXPORT_ARGUMENTS})
+
+  include("${rapids-cmake-dir}/cpm/detail/package_details.cmake")
+  rapids_cpm_package_details(bs_thread_pool version repository tag shallow exclude)
+
+  include("${rapids-cmake-dir}/cpm/detail/generate_patch_command.cmake")
+  rapids_cpm_generate_patch_command(bs_thread_pool ${version} patch_command)
+
+  include("${rapids-cmake-dir}/cpm/find.cmake")
+  rapids_cpm_find(bs_thread_pool ${version} ${ARGN}
+                  CPM_ARGS
+                  GIT_REPOSITORY ${repository}
+                  GIT_TAG ${tag}
+                  GIT_SHALLOW ${shallow} ${patch_command}
+                  EXCLUDE_FROM_ALL ${exclude}
+                  DOWNLOAD_ONLY ON)
+
+  include("${rapids-cmake-dir}/cpm/detail/display_patch_status.cmake")
+  rapids_cpm_display_patch_status(bs_thread_pool)
+
+  # Extract the major version value of bs_thread_pool to use with `rapids-export` to setup
+  # compatibility rules
+  include("${rapids-cmake-dir}/cmake/parse_version.cmake")
+  rapids_cmake_parse_version(MAJOR ${version} ${version})
+
+  # Set up install rules Need to be re-entrant safe so only call when `bs_thread_pool_ADDED`
+  if(bs_thread_pool_ADDED)
+    if(NOT TARGET thread_pool)
+      add_library(thread_pool INTERFACE)
+      target_include_directories(thread_pool
+                                 INTERFACE "$<BUILD_INTERFACE:${bs_thread_pool_SOURCE_DIR}/include>"
```

Comment:
We might want to enable `BS_THREAD_POOL_ENABLE_PAUSE` for all RAPIDS uses since it is enabled by cudf ( https://github.com/rapidsai/cudf/blob/branch-24.08/cpp/cmake/thirdparty/get_thread_pool.cmake )